File: resources/patterns/design/refresh-ahead-cache-pattern/refresh_ahead_cache.ts
// Refresh-ahead cache pattern implementation in TypeScript.
// Requires: redis@^4.0, TypeScript 5+
// Usage: ts-node refresh_ahead_cache.ts

import { createClient, RedisClientType } from 'redis';

interface RefreshConfig {
  loader: () => Promise<unknown>;
  ttl: number;
}

class RefreshAheadCache {
  private redis: RedisClientType;
  private refreshCallbacks: Map<string, RefreshConfig> = new Map();
  private refreshThreshold: number;
  private refreshTimer: NodeJS.Timeout | null = null;
  private running: boolean = true;

  constructor(redisClient: RedisClientType, refreshThreshold = 0.8) {
    this.redis = redisClient;
    this.refreshThreshold = refreshThreshold;
    this.startRefreshLoop();
  }

  async register(key: string, loader: () => Promise<unknown>, ttl: number): Promise<void> {
    this.refreshCallbacks.set(key, { loader, ttl });
    await this.refreshKey(key);
  }

  async get<T>(key: string): Promise<T | null> {
    const cached = await this.redis.get(key);
    if (cached !== null) {
      return JSON.parse(cached) as T;
    }
    const config = this.refreshCallbacks.get(key);
    if (config) {
      return (await this.refreshKey(key)) as T;
    }
    return null;
  }

  private startRefreshLoop(): void {
    this.refreshTimer = setInterval(() => this.checkAndRefresh(), 10000);
  }

  private async checkAndRefresh(): Promise<void> {
    for (const [key, config] of this.refreshCallbacks) {
      const ttlRemaining = await this.redis.ttl(key);
      if (ttlRemaining < 0 || ttlRemaining < config.ttl * (1 - this.refreshThreshold)) {
        try {
          await this.refreshKey(key);
        } catch (error) {
          console.error(`Refresh failed for ${key}:`, error);
        }
      }
    }
  }

  private async refreshKey(key: string): Promise<unknown> {
    const config = this.refreshCallbacks.get(key)!;
    const value = await config.loader();
    await this.redis.set(key, JSON.stringify(value), { EX: config.ttl });
    return value;
  }

  async shutdown(): Promise<void> {
    this.running = false;
    if (this.refreshTimer) clearInterval(this.refreshTimer);
  }
}

// Example usage
async function main() {
  const redis = createClient({ url: 'redis://localhost:6379' });
  await redis.connect();

  const cache = new RefreshAheadCache(redis, 0.8);

  await cache.register(
    'product:featured',
    async () => [{ id: 1, name: 'Widget' }],
    300
  );

  const products = await cache.get<{ id: number; name: string }[]>('product:featured');
  console.log('Featured products:', products);

  await cache.shutdown();
  await redis.quit();
}

main().catch(console.error);

File: resources/patterns/design/refresh-ahead-cache-pattern/refresh_ahead_cache.py
"""Refresh-ahead cache pattern implementation in Python.

Requires: redis>=4.0, Python 3.9+
Usage: python refresh_ahead_cache.py
"""

import json
import threading
import time
from typing import Any, Callable, Optional

import redis


class RefreshAheadCache:
    """Proactively refreshes cache entries before TTL expires."""

    def __init__(
        self,
        redis_client: redis.Redis,
        ttl: int = 3600,
        refresh_threshold: float = 0.8,
        check_interval: int = 10,
    ):
        self.redis = redis_client
        self.ttl = ttl
        self.refresh_threshold = refresh_threshold
        self.check_interval = check_interval
        self._refresh_callbacks: dict[str, dict] = {}
        self._running = True
        self._refresh_thread = threading.Thread(target=self._refresh_loop, daemon=True)
        self._refresh_thread.start()

    def register(self, key: str, loader: Callable[[], Any], ttl: Optional[int] = None) -> None:
        """Register a hot key for proactive refresh."""
        self._refresh_callbacks[key] = {"loader": loader, "ttl": ttl or self.ttl}
        self._refresh_key(key)

    def get(self, key: str) -> Any:
        """Get a value from cache. Falls back to synchronous load if missing."""
        cached = self.redis.get(key)
        if cached is not None:
            return json.loads(cached)
        if key in self._refresh_callbacks:
            return self._refresh_key(key)
        return None

    def _refresh_loop(self) -> None:
        """Background loop that checks and refreshes entries approaching TTL."""
        while self._running:
            time.sleep(self.check_interval)
            for key, config in list(self._refresh_callbacks.items()):
                ttl_remaining = self.redis.ttl(key)
                threshold = config["ttl"] * (1 - self.refresh_threshold)
                if ttl_remaining < 0 or ttl_remaining < threshold:
                    try:
                        self._refresh_key(key)
                    except Exception as e:
                        print(f"Refresh failed for {key}: {e}")

    def _refresh_key(self, key: str) -> Any:
        """Load data from the loader and update the cache."""
        config = self._refresh_callbacks[key]
        value = config["loader"]()
        self.redis.setex(key, config["ttl"], json.dumps(value))
        return value

    def shutdown(self) -> None:
        """Stop the background refresh thread."""
        self._running = False
        self._refresh_thread.join(timeout=10)


if __name__ == "__main__":
    # Example usage
    r = redis.Redis(host="localhost", port=6379, decode_responses=True)
    cache = RefreshAheadCache(r, ttl=300, refresh_threshold=0.8)

    # Register hot keys
    cache.register(
        "product:featured",
        lambda: [{"id": 1, "name": "Widget"}],
        ttl=300,
    )

    # Reads always hit cache
    products = cache.get("product:featured")
    print(f"Featured products: {products}")

    cache.shutdown()
